apollo-router/src/plugins/telemetry/config_new/cost/mod.rs

```diff
@@ -0,0 +1,358 @@
+use std::sync::Arc;
+
+use opentelemetry::metrics::MeterProvider;
+use opentelemetry_api::KeyValue;
+use parking_lot::Mutex;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use tower::BoxError;
+
+use crate::metrics;
+use crate::plugins::demand_control::CostResult;
+use crate::plugins::telemetry::config_new::attributes::SupergraphAttributes;
+use crate::plugins::telemetry::config_new::conditions::Condition;
+use crate::plugins::telemetry::config_new::extendable::Extendable;
+use crate::plugins::telemetry::config_new::instruments::CustomHistogram;
+use crate::plugins::telemetry::config_new::instruments::CustomHistogramInner;
+use crate::plugins::telemetry::config_new::instruments::DefaultedStandardInstrument;
+use crate::plugins::telemetry::config_new::instruments::Increment::Unit;
+use crate::plugins::telemetry::config_new::instruments::Instrumented;
+use crate::plugins::telemetry::config_new::selectors::SupergraphSelector;
+use crate::plugins::telemetry::config_new::Selectors;
+use crate::services::supergraph;
+use crate::services::supergraph::Request;
+use crate::services::supergraph::Response;
+use crate::Context;
+
+static COST_ESTIMATED: &str = "cost.estimated";
+static COST_ACTUAL: &str = "cost.actual";
+static COST_DELTA: &str = "cost.delta";
+
+/// Attributes for Cost
+#[derive(Deserialize, JsonSchema, Clone, Default, Debug, PartialEq)]
+#[serde(deny_unknown_fields, default)]
+pub(crate) struct SupergraphCostAttributes {
+    /// The estimated cost of the operation using the currently configured cost model
+    #[serde(rename = "cost.estimated")]
+    cost_estimated: Option<bool>,
+    /// The actual cost of the operation using the currently configured cost model
+    #[serde(rename = "cost.actual")]
+    cost_actual: Option<bool>,
+    /// The delta (estimated - actual) cost of the operation using the currently configured cost model
+    #[serde(rename = "cost.delta")]
+    cost_delta: Option<bool>,
+    /// The cost result, this is an error code returned by the cost calculation or COST_OK
+    #[serde(rename = "cost.result")]
+    cost_result: Option<bool>,
+}
+
+impl Selectors for SupergraphCostAttributes {
+    type Request = supergraph::Request;
+    type Response = supergraph::Response;
+
+    fn on_request(&self, _request: &Self::Request) -> Vec<KeyValue> {
+        Vec::default()
+    }
+
+    fn on_response(&self, response: &Self::Response) -> Vec<KeyValue> {
+        let mut attrs = Vec::with_capacity(4);
+
+        if let Some(true) = self.cost_estimated {
+            if let Some(cost_result) = &response.context.extensions().lock().get::<CostResult>() {
```

Comment:
Is this lock statement repeated just to keep the critical sections small?

Reply:
Oops, rookie mistake: [c9ca854](https://github.com/apollographql/router/pull/4990/commits/c9ca854a6837c890b19fee0e2c120ecd68802120)